### ISSUE TITLE
False positive on `CanBeNonNullable`: not differentiating between `&&` and `||` in some scenarios

### ISSUE THREAD

**Issue by eygraber** [detekt/detekt#5799]
If my code looks like this, I get the false positive:

```kotlin
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.consumeWindowInsets
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier

data class ScaffoldState(
  val appBarTitle: String? = null
)

@OptIn(ExperimentalLayoutApi::class, ExperimentalMaterial3Api::class)
@Composable
fun DetektReproScaffold(
  state: ScaffoldState,
  shouldNavigate: Boolean,
  modifier: Modifier = Modifier
) {
  Scaffold(
    modifier = modifier,
    topBar = {
      HtTopAppBar(
        shouldNavigate = shouldNavigate,
        title = state.appBarTitle
      )
    }
  ) { innerPadding ->
    Box(
      modifier = Modifier
        .consumeWindowInsets(innerPadding)
        .padding(innerPadding)
    )
  }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun HtTopAppBar(
  shouldNavigate: Boolean,
  modifier: Modifier = Modifier,
  title: String? = null
) {
  if(shouldNavigate || title != null) {
    TopAppBar(
      title = {
        if(title != null) {
          Text(
            text = title
          )
        }
      },
      modifier = modifier
    )
  }
}
```

However, if I move the if outside of the function then the false positive does not occur:

```kotlin
@OptIn(ExperimentalLayoutApi::class, ExperimentalMaterial3Api::class)
@Composable
fun DetektReproScaffold(
  state: ScaffoldState,
  shouldNavigate: Boolean,
  modifier: Modifier = Modifier
) {
  Scaffold(
    modifier = modifier,
    topBar = {
      if(shouldNavigate || state.appBarTitle != null) {
        HtTopAppBar(
          title = state.appBarTitle
        )
      }
    }
  ) { innerPadding ->
    Box(
      modifier = Modifier
        .consumeWindowInsets(innerPadding)
        .padding(innerPadding)
    )
  }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun HtTopAppBar(
  modifier: Modifier = Modifier,
  title: String? = null
) {
  TopAppBar(
    title = {
      if(title != null) {
        Text(
          text = title
        )
      }
    },
    modifier = modifier
  )
}
```

**Comment by marschwar** [comment]
So basically what you are saying is. The null check is not the only thing that is being evaluated and therefor the function does actually do something even when the value is null. Just like in the example below. 

```kotlin
fun foo(a: Int?, other: Int) {
    if (a != null || other % 2 == 0) {
        println(a + 5)
    }
}
```

**Comment by eygraber** [comment]
Yes, but in my case if `a` is null we'd still do something, but wouldn't use `a`

**Comment by marschwar** [comment]
While looking into this I found that parenthesized expression are not handled correctly. There is even one test case that I think is verifying the wrong outcome.

https://github.com/detekt/detekt/blob/c3bb324b833f28046664b44be77941c57e76db39/detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableSpec.kt#L852-L866

In my opinion this should be reported as a violation. Whenever b is null, this method will not do anything. Am I wrong on this one? And how does the operator function come into play here?

**Comment by marschwar** [comment]
> Yes, but in my case if `a` is null we'd still do something, but wouldn't use `a`

Of course my example is nonsense. I updated it.

**Comment by atulgpt** [comment]
> While looking into this I found that parenthesized expression are not handled correctly. There is even one test case that I think is verifying the wrong outcome.
> 
> https://github.com/detekt/detekt/blob/c3bb324b833f28046664b44be77941c57e76db39/detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableSpec.kt#L852-L866
> 
> In my opinion this should be reported as a violation. Whenever b is null, this method will not do anything. Am I wrong on this one? And how does the operator function come into play here?

Hi, @marschwar this seems to be a wrong assertion.

**Comment by atulgpt** [comment]
I think the baseline is that whenever we have `!=` check with `||` we can assume that the nullability of a value is being used while in the case of `!=` check with `&&` we are not doing anything with nullability value. Are there any more cases that should be handled?
